hack/scripts/provision.sh

```diff
@@ -0,0 +1,996 @@
+create_physical_volume() {
+	local disk_path="$1"
+
+	# if already exists, do nothing
+	if [[ $(pvdisplay 2>/dev/null) != *"$disk_path"* ]]; then
+		pvcreate -q "$disk_path" || die "failed to create physical volume on $disk_path"
+		say "Created physical volume on $disk_path"
+		return 0
+	fi
+
+	say "Physical volume on $disk_path already exists"
+}
+
+# Create a volume group on the given device for the thinpool
+create_volume_group() {
+	local disk_path="$1"
+	local thinpool="$2"
+
+	# if already exists, do nothing
+	if [[ $(vgdisplay 2>/dev/null) != *"$thinpool"* ]]; then
+		vgcreate -q "$thinpool" "$disk_path" || die "failed to create volume group on $disk_path"
+		say "Created volume group on $disk_path"
+		return 0
+	fi
+
+	say "Volume group on $disk_path already exists"
+}
+
+# Format the volume for thinpool storage
+create_logical_volume() {
+	local volume_group="$1"
+
+	# if already exists, do nothing
+	if [[ $(lvdisplay 2>/dev/null) != *"$volume_group"* ]]; then
+		lvcreate -q --wipesignatures y -n thinpool "$volume_group" -l 95%VG || die "Failed to create logical volume for thinpool data"
+		say "Created logical volume for $volume_group thinpool data"
+
+		lvcreate -q --wipesignatures y -n thinpoolmeta "$thinpool" -l 1%VG || die "Failed to create logical volume for thinpool metadata"
+		say "Created logical volume for $volume_group thinpool metadata"
+
+		lvconvert -q -y \
+			--zero n \
+			-c 512K \
+			--thinpool "$volume_group"/thinpool \
+			--poolmetadata "$volume_group"/thinpoolmeta || die "Failed to convert logical volumes to thinpool storage"
+		say "Converted logical volumes to thinpool storage"
+
+		return 0
+	fi
+
+	say "Logical volume for $volume_group thinpool already exists"
+}
+
+# Create and apply the lvm profile for the thinpool
+apply_lvm_profile() {
+	local thinpool="$1"
+	local profile="$THINPOOL_PROFILE_PATH/$thinpool-thinpool.profile"
+
+	if [[ ! -f "$profile" ]]; then
+		cat <<'EOF' >>"$profile"
+activation {
+thin_pool_autoextend_threshold=80
+thin_pool_autoextend_percent=20
+}
+EOF
+		say "Written lvm profile to $profile"
+	fi
+
+	# if already exists, do nothing
+	if [[ $(lvs 2>/dev/null) != *"$thinpool"* ]]; then
+		lvchange -q --metadataprofile "$thinpool-thinpool" "$thinpool"/thinpool || die "Could not apply lvm profile $profile"
+		say "Applied lvm profile for $thinpool-thinpool"
+		return 0
+	fi
+
+	say "LVM profile for $thinpool-thinpool already applied"
+}
+
+# Try 5 times to ensure the lvm profile is monitored
+monitor_lvm_profile() {
+	local thinpool="$1"
+
+	try=1
+	max=5
+	while [ "$try" -le "$max" ]; do
+		say "Checking that lvm profile for $thinpool-thinpool is monitored"
+		if [[ $(lvs -o+seg_monitor 2>/dev/null) == *"not monitored"* ]]; then
+			lvchange --monitor y "$thinpool"/thinpool
+			((try = try + 1))
+
+			if [[ "$try" -gt "$max" ]]; then
+				exit 1
+			fi
+
+			continue
+		fi
+
+		break
+	done
+
+	say "Successfully monitored $thinpool-thinpool profile"
```

Comment:
```suggestion
  for counter in $(seq 5); do
    if [[ $(lvs -o+seg_monitor 2>/dev/null) != *"not monitored"* ]]; then
      say "Successfully monitored ${thinpool}-thinpool profile"

      return
    fi

    lvchange --monitor y "${thinpool}/thinpool"
  done

  die -c 1
```

I think it does the same but a less complex.

Reply:
nice!